src/BigQuery/BigQueryClient.php

```diff
@@ -0,0 +1,355 @@
+<?php
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\BigQuery;
+
+use Google\Cloud\BigQuery\Connection\ConnectionInterface;
+use Google\Cloud\BigQuery\Connection\Rest;
+
+/**
+ * Google Cloud BigQuery client. Allows you to create, manage, share and query
+ * data. Find more information at
+ * [Google Cloud BigQuery Docs](https://cloud.google.com/bigquery/what-is-bigquery).
+ */
+class BigQueryClient
+{
+    use JobConfigurationTrait;
+
+    const SCOPE = 'https://www.googleapis.com/auth/bigquery';
+    const INSERT_SCOPE = 'https://www.googleapis.com/auth/bigquery.insertdata';
+
+    /**
+     * @var ConnectionInterface $connection Represents a connection to BigQuery.
+     */
+    protected $connection;
+
+    /**
+     * @var string The project ID created in the Google Developers Console.
+     */
+    private $projectId;
+
+    /**
+     * Create a BigQuery client.
+     *
+     * Example:
+     * ```
+     * use Google\Cloud\BigQuery\BigQueryClient;
+     *
+     * $bigQuery = new BigQueryClient([
+     *     'projectId' => 'myAwesomeProject'
+     * ]);
+     * ```
+     *
+     * @param array $config {
+     *     Configuration options.
+     *
+     *     @type string $projectId The project ID from the Google Developer's
```

Comment:
PHP community has started to experiment with Builder pattern, also started to think type hinting is a nice thing.

What's your general thoughts on using Builder pattern instead of $config array?
In general, I think IDEs can support you better with the Builder pattern. Of course, using $config array is more conventional at this point, but I feel like the Builder pattern will likely soon become very common too. I may be wrong.